pkg/reconcilermanager/controllers/otel_controller.go

```diff
@@ -131,11 +132,16 @@ func (r *OtelReconciler) reconcileConfigMap(ctx context.Context, req reconcile.R
 // configureGooglecloudConfigMap creates or updates a map with a config that
 // enables Googlecloud exporter if Application Default Credentials are present.
 func (r *OtelReconciler) configureGooglecloudConfigMap(ctx context.Context) ([]byte, error) {
-	// Check that GCP credentials are injected
-	creds, _ := getDefaultCredentials(ctx)
-	if creds == nil || creds.ProjectID == "" {
-		// No injected credentials
-		return nil, nil
+	// Only configure otel-collector to export metrics to Cloud Monitoring if
+	// GCP credentials are configured on the reconciler-manager.
+	// This assumes the otel-collector will be similarly configured.
+	_, err := r.credentialProvider.Credentials()
+	if err != nil {
+		if auth.IsCredentialsNotFoundError(err) {
```

Comment:
Can we continue using `creds == nil || creds.projectID == ""` to check the existence of injected credentials.

Using error message is error-prone when updating the otel-collector version.

Reply:
There's a couple problems with that...
1. The CredentialProvider interface I added actually returns a TokenProvider, not a full Credentials object. I did this, because Credentials is a struct, not an interface, and we only really needed the Token method anyway. Plus using TokenProvider simplifies the fake for the test.
2. Even if we changed the CredentialProvider to return a Credentials struct pointer, or a new custom interface, the Credentials struct doesn't expose the ProjectID as a public field any more. Instead, it has a `ProjectID(context.Context) (string, error)` method, which uses the `CredentialsPropertyProvider` interface to get the project ID.  https://github.com/googleapis/google-cloud-go/blob/auth/v0.12.0/auth/auth.go#L159

AFAICT, the project ID is generally parsed from the json, but the `CredentialsPropertyProvider` interface allows the project ID to be injected, overriding the `project_id` or `project` in the json. To add extra complexity, it looks like GDCH uses a `project` field, instead of a `project_id` filed. So the `project` field value is used if the `project_id` is missing or empty.
https://github.com/googleapis/google-cloud-go/blob/auth/v0.12.0/auth/internal/internal.go#L121

Then there's the `internal.GetProjectID` method, which allows overriding the project ID with the `GOOGLE_CLOUD_PROJECT` env var. So if that's set, checking the `Credentials.ProjectID` will always return the value of the env var, telling us nothing about whether the credentials json is valid.

So, I could add a new Credentials interface that exposes the ProjectID method, update the fake to impl it, and make the reconciler-manager check that ProjectID doesn't error and is non-empty, but I'm not entirely sure how to handle errors from the ProjectID method. The `CredentialsPropertyProvider.GetProperty` method doesn't seem to have any sort of contract as to what kind of errors it might return. Should they be ignored? Should they be terminal?

Reply:
Anyway, the impl in this PR will surface all other errors from `Credentials()`, which seems like a good improvement. Previously, when the config was invalid it would just silently skip creating the ConfigMap, defaulting otel-collector to just the Kubernetes+Prometheus exporters.

Reply:
I can add a test for IsCredentialsNotFoundError tho that will error if the error text is changed upstream.

Reply:
> I can add a test for IsCredentialsNotFoundError tho that will error if the error text is changed upstream.

That works.

Reply:
Added. It does make the tests more fragile, but that's better than fragile production code, i guess.